### PR TITLE
Add option for exact inference (for single factor gen. models) w/ associative scan; online learning w/ sequential inference methods (MMP, VMP, OVF, exact)

### PR DESCRIPTION
**Pull request by conorheins**
## Summary
This PR adds an **`exact` inference algorithm** for single-factor HMMs using associative scans (prefix scan for filtering, suffix scan for smoothing), and integrates it through `Agent`, `infer_states`, and `rollout`.

It also introduces a fixed-shape rollout path for windowed sequence inference/learning so there is a single compile-time shape, which also preserves warmup/sequence semantics across `mmp`, `vmp`, `ovf`, and `exact` algos.

### Exact inference integration
- Added `exact` inference for single-factor HMMs in `pymdp/algos.py` with associative scans:
  - Filtering (`hmm_filter_scan_rowstoch`, `hmm_filter_scan_colstoch`)
  - Filtering + smoothing (`hmm_smoother_scan_rowstoch`, `hmm_smoother_scan_colstoch`)
  - Smoothing from precomputed filtered marginals (`hmm_smoother_from_filtered_colstoch`)
- Integrated exact scan routing through:
  - `pymdp/inference.py` (`ONE_STEP_METHODS`, `SEQUENCE_METHODS`, `SMOOTHING_METHODS`)
  - `Agent` config via `inference_algo="exact"`
  - `run_exact_single_factor_hmm_scan(...)`
  - rollout learning path for smoothing-based parameter updates

### Fixed-shape rollout execution (no `lax.switch` per `k`)
- Refactored `pymdp/envs/rollout.py` to keep fixed-size buffers for:
  - observations
  - actions
  - filtered posterior history (`qs`)
- Replaced growing-length window branching with fixed-shape execution and a `valid_steps` mask.
- Preserved warmup behavior and rolling-window semantics for all sequence/smoothing methods.
- Kept `MAX_WINDOWED_HISTORY_WITHOUT_HORIZON` guard for uncapped windowing with warning.

### Window-aware inference/learning plumbing
- Extended `Agent.infer_states(...)` with optional `valid_steps` and threaded it through to inference.
- Extended `update_posterior_states(...)` with `valid_steps` for sequence methods.
- Added mask-aware transition handling in `run_mmp`/`run_vmp` message passing (including sequence-validity masking and no-op boundaries).
- Added masked transition handling in inference/learning flow so padded timesteps do not alter valid posterior updates.
- For sequence methods, empirical priors remain frozen during warmup and then shift forward via the dropped oldest action once full/rolling.

### Smoothing and learning robustness
- `smoothing_ovf(...)` now handles padded/invalid (negative) actions safely in sparse path transitions.
- Invalid padded transitions are explicitly no-ops/identity where needed, preventing contamination in warmup smoothing windows.
- `infer_parameters(...)` now allows A-updates from smooth states even when `learn_B=False`, while B-updates remain gated by `learn_B`/valid-transition availability.
- B-update gating now requires actual valid transitions, not merely that a valid action array exists.

### API cleanup and de-duplication
- Simplified `Agent.update_empirical_prior(...)` to return only `pred` (prior), removing tuple return of `(pred, qs)` and updating callsites/tests.
- Removed the backwards-compatible `joint_dist_factor_conditioned(...)` helper and now use `joint_dist_factor(..., actions=None)` directly where conditioned transitions are precomputed.
- Restored detailed docstrings/comments in `pymdp/algos.py` for HMM scan entry points for future maintainability.

## Tests
- Added/updated rollout-vs-manual-reference regression coverage for:
  - fixed-shape vs warmup semantics for `mmp`, `vmp`, `ovf`, `exact`
  - online parameter updates with windowed histories
  - sequence/smoothing parity under warmup and sliding windows
- Added/update masking and padded-transition tests for OVF sparse invalid-action handling.
- Updated agent/JAX and inference tests for updated API behavior and prior-return shape.
- Kept exact-associative test coverage for:
  - row/col convention consistency
  - scan-vs-reference correctness
  - sequence alignment and masking behavior
  - gradient finiteness edge cases

### Validation runs
- `pytest test/test_rollout_function.py test/test_hmm_associative_scan.py test/test_learning_jax.py test/test_agent_jax.py -q`

## Notebook updates
Updated notebook callsites to match the `update_empirical_prior` API cleanup:
- `examples/inference_and_learning/inference_methods_comparison.ipynb`
- `examples/sparse/sparse_benchmark.ipynb`
- `examples/advanced/complex_action_dependency.ipynb`

### CODE REVIEW

File: pymdp/algos.py [hunk 1]
```diff
@@ -465,6 +470,529 @@ def scan_fn(carry, t):
...
+    if transition_mats.ndim == 2:
+        transition_mats = jnp.broadcast_to(transition_mats, (T-1, K, K))
+
+    @vmap
+    def _marginalize(m_ij: FilterMessage, m_jk: FilterMessage) -> FilterMessage:
+        # Combine segment i->j with segment j->k by summing out z_j
+        A_ij_cond, lognorm = _condition_on(m_ij.A, m_jk.log_b)
+        A_ik = A_ij_cond @ m_jk.A
+        log_b_ik = m_ij.log_b + lognorm
+        return FilterMessage(A=A_ik, log_b=log_b_ik)
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/algos.py:661]: I think these comments are not correct for this code (both row- and column-stochastic cases).

In `lax.associative_scan`, the combiner is called on batched chunks along the scan axis, not just single (K,K)/(K) elements.
Here, `_marginalize` is intentionally @vmap’d to process that chunk dimension.

File: pymdp/algos.py [hunk 2]
```diff
@@ -465,6 +470,529 @@ def scan_fn(carry, t):
...
+    if B_mats.ndim == 2:
+        B_mats = jnp.broadcast_to(B_mats, (T - 1, K, K))
+
+    @vmap
+    def _marginalize(m_ij: FilterMessage, m_jk: FilterMessage) -> FilterMessage:
+        # Combine segment i->j with segment j->k by summing out z_j (end-state axis)
+        A_ij_cond, lognorm = _condition_on(m_ij.A, m_jk.log_b, axis=0)
+        A_ik = m_jk.A @ A_ij_cond
+        log_b_ik = m_ij.log_b + lognorm
+        return FilterMessage(A=A_ik, log_b=log_b_ik)
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/algos.py]: I think these comments are not correct for this code (both row- and column-stochastic cases).

In `lax.associative_scan`, the combiner is called on batched chunks along the scan axis, not just single (K,K)/(K) elements.
Here, `_marginalize` is intentionally @vmap’d to process that chunk dimension.

File: pymdp/envs/rollout.py [hunk 3]
```diff
@@ -140,30 +256,164 @@ def step_fn(carry, t):
...
+                        learning_outcomes=obs_k,
+                        learning_actions=actions_k,
+                        learning_beliefs=qs_prev_k,
+                    )
+                    qs_latest_k = jtu.tree_map(lambda x: x[:, -1, ...], qs_k)
+                    return updated_agent_k, action_next_k, qs_k, qs_latest_k, xtra_k
+
+                return _branch
+
+            branches = tuple(_window_branch(k) for k in range(1, history_len + 1))
```
Comment (conorheins) [inline pymdp/envs/rollout.py]: @praesc raised a good point here (please feel free to elaborate if i misrepresent this at all) -- but having each of these branches for a fixed window of size `k = 1, 2, ..., inference_horizon-1` is going to lead to a bunch of differently compiled versions of `infer_and_plan`, which in context of large values of `inference_horizon`, will burden memory. Ideally, we can do this with padding instead and fixed sizes of observation, action and qs_prev history

Comment (praesc) [reply pymdp/envs/rollout.py]: Correct. My findings were that k changes the posterior shape, and a different kernel is compiled & invoked for each shape. This affects both JITting latency (compiling len(inference_horizon) kernels instead of one) and inference latency (switching kernel context for t<infer_horizon). 
It would also affect memory, as we'll have several kernels in memory instead of 1. XLA might eventually dump the kernels to disk and free some RAM, but it's an extra memory we don't have to incur.

Comment (conorheins) [reply pymdp/envs/rollout.py]: Okay, have now tried to address this with f2d2159 -- no more switches between different k-length JITted horizon paths, now everything uses the same inference_horizon-length path (one jitted function), but with padding for early timesteps. The downside of this was that I now have to thread a bunch of `valid_steps` masks down through the callsites of `inference.py` and `algos.py`, but I think that is worth it -- because this sort of masking/padding is something that will be generically useful IMO

File: pymdp/inference.py [hunk 4]
```diff
@@ -127,19 +281,103 @@ def step_fn(qs_smooth, xs):
...
+    if any(isinstance(b, JAXSparse) for b in B):
+        joint = lambda b, qs, f: joint_dist_factor(b, qs, past_actions[..., f])
+        marginals_and_joints = ([], [])
+        for b, qs, f in zip(B, filtered_post, list(range(nf))):
+            marginals, joints = joint(b, qs, f)
+            marginals_and_joints[0].append(marginals)
+            marginals_and_joints[1].append(joints)
+    else:
+        B_seq = _condition_transitions_on_actions(B, past_actions, invalid_action_mode="identity")
+        joint = lambda b_seq, qs: joint_dist_factor_conditioned(b_seq, qs)
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/inference.py]: Hopefully this has now been addressed by 8fe7c93

File: pymdp/agent.py [hunk 5]
```diff
@@ -436,50 +443,89 @@ def infer_parameters(self, beliefs_A, outcomes, actions, beliefs_B=None, lr_pA=1
...
         # A handy predicate: can we (meaningfully) update B now?
-        # We need at least one transition and all actions to be non-negative (sentinel -1 = invalid).
+        # We need at least one valid transition.
         can_update_B = (
             self.learn_B
             and actions_Tm1 is not None
             and (T > 1)
             and (actions_Tm1.shape[1] == (T - 1))
-            and jnp.all(actions_Tm1 >= 0)
+            and jnp.any(valid_transition_mask)
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/agent.py]: This should now have been addressed by [b6ae844](https://github.com/infer-actively/pymdp/pull/349/commits/b6ae8440bdef070264fdc6039f776f6624bf25e2) and [4d75fc9](https://github.com/infer-actively/pymdp/pull/349/commits/4d75fc91b5997b80226b86ca187126259186efc5)

File: pymdp/algos.py [hunk 6]
```diff
@@ -465,6 +557,472 @@ def scan_fn(carry, t):
...
+
+    mll, q_filt, q_pred, q_smooth, q_t_given_next = hmm_smoother_scan_colstoch(pi, B_seq, ll)
+
+    # Return in pymdp-ish packaging (list per factor)
+    qs = [q_smooth]
+    ps = [q_pred]
+    qss = [q_t_given_next]   # shape (T-1, K_next, K_curr)
+
+    return mll, qs, ps, qss
+
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/algos.py:1024]: This is used for internal validation

File: pymdp/envs/rollout.py [hunk 7]
```diff
@@ -17,39 +54,330 @@ def default_policy_search(agent, qs, rng_key):
...
+def _init_windowed_carry(agent, observation_0, env_state, rng_key, history_len, action_history_len, is_sequence_method):
+    action_0 = -jnp.ones((agent.batch_size, agent.policies.policy_arr.shape[-1]), dtype=jnp.int32)
+
+    qs_0 = jtu.tree_map(
+        lambda x: jnp.broadcast_to(
+            jnp.expand_dims(x, axis=1),
+            (x.shape[0], history_len, x.shape[-1]),
+        ),
+        agent.D,
+    )
```
[the remark this answers is not in the document]
Comment (conorheins) [reply pymdp/envs/rollout.py:300]: You’re right that the buffers are initialized differently - observations are right-aligned with padded prefix, while `qs_0` is initialized with D across the full window. 

In this rollout path that is intentional, because sequence inference only uses the valid suffix via `valid_steps`, so padded-prefix observation/belief slots are masked out. Transition terms are also masked by action validity (initial padded actions are -1), so padded transitions do not contribute. As `valid_steps` grows, those slots become valid and are overwritten through the rolling window updates.

So this is not a correctness issue in the current implementation, but your point is valid from a readability/defensive-programming perspective. I’ve added an inline comment near `qs_0`init to make this explicit.
